core/trino-main/src/main/java/io/trino/operator/TopNOperator.java

```diff
@@ -204,22 +230,27 @@ private void addPage(Page page)
         @Override
         public TransformationState<Page> process(Page inputPage)
         {
-            if (inputPage != null) {
-                addPage(inputPage);
-                return TransformationState.needsMoreData();
-            }
+            checkState(inputPage == null || isPartialFlushing, "Input page is not null without partial flushing");
 
-            // no more input, return results
-            Page page = null;
-            while (page == null && !topNProcessor.noMoreOutput()) {
-                page = topNProcessor.getOutput();
-            }
+            Page page = topNProcessor.getOutput();
 
             if (page != null) {
-                return TransformationState.ofResult(page, false);
+                if (isPartialFlushing) {
+                    pageBuffer.add(inputPage); // in order to keep calling TopNPages.process() until partial flushing is completed
```

Comment:
@sopel39, this looks like an awkward way to do this. Do you have any suggestions on the right structure for this pattern?

Reply:
I don't like this approach, either. This is like picking the lesser evil. @sopel39 if you have a better way to accomplish this, please let me know.

Reply:
I changed the implementation to add one more method in Transformation interface. With this, operators implementing WorkProcessor can handle input buffer full situation.